tool-plugins/nextjs-starter/src/hooks/useAppBridge.ts

```diff
@@ -0,0 +1,277 @@
+import {
+	AppBridgeSession,
+	BeginOAuthMessagePayload,
+	CreateBeginOAuthMessagePayload,
+	CreateValidateMessagePayload,
+	PluginType,
+	PostMessageAction,
+	ValidateMessagePayload,
+} from '@/types';
+import {
+	APP_BRIDGE_ORIGIN,
+	KEY_PARENT_HOST,
+	KEY_SLUG,
+	KEY_TOKEN,
+	KEY_VALIDATED_PAYLOAD,
+} from '@/utils/const';
+import { useState, useEffect } from 'react';
+
+const getPostMessageAction = (type: PluginType): PostMessageAction => {
+	switch (type) {
+		case 'space-plugin':
+			return 'app-changed';
+		case 'tool-plugin':
+			return 'tool-changed';
+		default:
+			throw new Error(`Invalid plugin type: ${type}`);
+	}
+};
+
+const getParentHost = () => {
+	const storedHost = sessionStorage.getItem(KEY_PARENT_HOST);
+	if (storedHost) {
+		return storedHost;
+	}
+	const params = new URLSearchParams(location.search);
+	const protocol = params.get('protocol');
+	const host = params.get('host');
+	if (!protocol || !host) {
+		throw new Error('Missing `protocol` or `host` in query params');
+	}
+	return `${protocol}//${host}`;
+};
+
+const getSlug = () => {
+	const storedSlug = sessionStorage.getItem(KEY_SLUG);
+	if (storedSlug) {
+		return storedSlug;
+	}
+	const params = new URLSearchParams(location.search);
+	return params.get('slug');
+};
+
+const postMessageToParent = (payload: unknown) => {
+	window.parent.postMessage(payload, getParentHost());
+};
+
+const useAppBridgeAuth = ({
+	type,
+	authenticated,
+}: {
+	type: PluginType;
+	authenticated: () => Promise<void>;
+}) => {
+	const [status, setStatus] = useState<
+		'init' | 'authenticating' | 'authenticated' | 'error'
+	>('init');
+	const [error, setError] = useState<unknown>();
+
+	const init = async () => {
+		const isInIframe = window.top !== window.self;
+
+		if (!isInIframe) {
+			setStatus('error');
+			setError('not-in-iframe');
+			return;
+		}
+
+		if (!isAuthenticated()) {
+			sendValidateMessageToParent();
+			return;
+		}
+
+		setStatus('authenticated');
+		setError(undefined);
+
+		await authenticated();
+	};
+
+	const isAuthenticated = () => {
+		try {
+			const payload: AppBridgeSession = JSON.parse(
+				sessionStorage.getItem(KEY_VALIDATED_PAYLOAD) || '',
+			);
+			return payload && new Date().getTime() / 1000 < payload.exp;
+		} catch (err) {
+			return false;
+		}
+	};
+
+	const sendValidateMessageToParent = () => {
+		setStatus('authenticating');
+		setError(undefined);
+		const host = getParentHost();
+		const slug = getSlug();
+
+		try {
+			const payload = createValidateMessagePayload({ type, slug });
+
+			postMessageToParent(payload);
+			sessionStorage.setItem(KEY_PARENT_HOST, host);
+			sessionStorage.setItem(KEY_SLUG, slug || '');
+		} catch (err) {
+			sessionStorage.removeItem(KEY_PARENT_HOST);
```

Comment:
Hey @eunjae-lee 🙌 

Is there any specific reason for not removing also the `KEY_SLUG`? 

I mean, if this method is called another time but fails before setting the session but succeeds in the previous attempt, then it may have both the `KEY_PARENT_HOST` and `KEY_SLUG` set in the previous successful call, so resetting both may be a good approach?!

I don't know if would be also nice to call the `setError(...)` for that case...

Reply:
I think you're right. I tried to remember if there was any reason for this, but I cannot. So I might've just missed it. I will apply the change.